[Oxidize BasisTranslator] Add rust-native `compose_transforms()`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits add the helper function `compose_transforms` from the `BasisTranslator` transpiler pass into rust. This is a small progression towards completing #12246. This function will be temporarily exposed to Python until #12246 is fully realized.


### Details and comments
As we continue moving parts of the `BasisTranslator` into Rust thanks to #12585 and #12811. The next logical step is to bring in those functions that relied heavily on the `DAGCircuit` and since it now lives in Rust we can now leverage its usage by moving this helper function into Rust.

`compose_transforms` basically returns a mapping of `(gate_name, num_qubits) : (parameters, gate_definition_as_dag)` for each gate present in the `DAGCircuit` that is currently being processed. It also breaks down `ControlFlowOp` to include its definitions as well.

Using the latest additions by #13036 we can perform most of these conversions from rust using both `circuit_to_dag` or `DAGCirctut::from_circuit_data`. This function will be temporarily exposed to Python until #12246 is fully realized.

### Blockers
- [x] #13143

**These are not strictly blockers but some nice-to-haves.**
- [x] #12585 
- [ ] #12811

raynelfss committed:
Initial: Add `compose_transforms` and `get_example_gates` to Rust.

raynelfss committed:
Add: `compose_transform` logic in rust

raynelfss committed:
Fix: Correct the behavior of `compose_transforms`.
-  Use `PackedOperation.control_flow` instead of `CONTROL_FLOW_OP_NAMES` to check if an instructuion is a control flow operation.
- Remove panic statement when checking for example gates.
- Use qreg when adding an instruction to the mock_dag in `compose_transforms`.
- Add missing comparison check in for loop to compare the mapped instructions.
- Use borrowed `DAGCircuit` instances in the recursion of `get_example_gates`, do not use extract.
- Fix behavior of `get_example_gates` to return `PackedInstruction` instances instead of `NodeIndex`.

raynelfss committed:
Merge main

raynelfss committed:
Fix: Leverage rust-based `circuit_to_dag` converters.
- Use `circuit_to_dag` and `DAGCircuit::from_circuit_data` to convert `QuantumCircuit` instances.

raynelfss committed:
Merge branch 'main' into rust-compose-trans

raynelfss committed:
Format: Fix indentation of import lines in `compose_transforms.rs`

raynelfss commented:
Here are some benchmarks
```bash
All benchmarks:

| Change   | Before [81063a79] <fix-abi3-issues~1>   | After [e8c957df] <rust-compose-trans>   |   Ratio | Benchmark (Parameter)                                                                                    |
|----------|-----------------------------------------|-----------------------------------------|---------|----------------------------------------------------------------------------------------------------------|
|          | 306±1ms                                 | 303±2ms                                 |    0.99 | passes.MultipleBasisPassBenchmarks.time_basis_translator(14, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
|          | 443±3ms                                 | 437±3ms                                 |    0.99 | passes.MultipleBasisPassBenchmarks.time_basis_translator(20, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
|          | 188±0.8ms                               | 182±2ms                                 |    0.97 | passes.MultipleBasisPassBenchmarks.time_basis_translator(14, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
|          | 217±0.7ms                               | 207±0.6ms                               |    0.96 | passes.MultipleBasisPassBenchmarks.time_basis_translator(20, 1024, ['u', 'cx', 'id'])                    |
|          | 102±0.5ms                               | 98.1±1ms                                |    0.96 | passes.MultipleBasisPassBenchmarks.time_basis_translator(5, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id'])  |
|          | 266±4ms                                 | 253±0.5ms                               |    0.95 | passes.MultipleBasisPassBenchmarks.time_basis_translator(20, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
|          | 155±3ms                                 | 146±0.6ms                               |    0.94 | passes.MultipleBasisPassBenchmarks.time_basis_translator(14, 1024, ['u', 'cx', 'id'])                    |
|          | 68.6±1ms                                | 62.6±0.2ms                              |    0.91 | passes.MultipleBasisPassBenchmarks.time_basis_translator(5, 1024, ['rz', 'x', 'sx', 'cx', 'id'])         |
| -        | 57.5±0.4ms                              | 51.8±1ms                                |    0.9  | passes.MultipleBasisPassBenchmarks.time_basis_translator(5, 1024, ['u', 'cx', 'id'])                     |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

```bash
All benchmarks:

| Change   | Before [81063a79] <fix-abi3-issues~1>   | After [e8c957df] <rust-compose-trans>   | Ratio   | Benchmark (Parameter)                                                                                     |
|----------|-----------------------------------------|-----------------------------------------|---------|-----------------------------------------------------------------------------------------------------------|
|          | 5.91±0s                                 | 5.13±0.01s                              | ~0.87   | randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile([0, 1])               |
|          | 5.93±0.02s                              | 5.11±0s                                 | ~0.86   | randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread([0, 1]) |
| -        | 2.53±0.01s                              | 2.26±0.01s                              | 0.89    | randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile([0])                  |
| -        | 2.56±0.01s                              | 2.25±0.01s                              | 0.88    | randomized_benchmarking.RandomizedBenchmarkingBenchmark.time_ibmq_backend_transpile_single_thread([0])    |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

raynelfss committed:
Formatting: Separate complicated type aliases

raynelfss committed:
Merge branch 'main' into rust-compose-trans

raynelfss committed:
Fix: Adapt to new `DAGCircuit` limitations

raynelfss committed:
Merge branch 'main' into rust-compose-trans

raynelfss committed:
Format: Remove unused imports in BasisTranslator

raynelfss committed:
Merge branch 'main' into rust-compose-trans

raynelfss committed:
Merge branch 'Qiskit:main' into rust-compose-trans

raynelfss committed:
Fix: Adapt to #13143

kevinhartman commented on crates/accelerate/src/basis/basis_translator/mod.rs:
```diff
@@ -0,0 +1,21 @@
+// This code is part of Qiskit.
```
Is there a clear benefit to having a separate folder here for `basis_translator`?

I don't know that I mind it, but it's different from the folder structure we had in Python, and none of the other accelerate modules really go more than one module deep (except for `synthesis`, which seems to copy the organization of the Python module).

raynelfss replied:
This is just to plan ahead, since the `BasisTranslator` has a couple of moving parts and since I am separating things at the moment to make them more readable. If you take a look at #12811, that feature also exists in that same folder. but in a different file If you think it shouldn't be this way it can be changed.

kevinhartman replied:
It suppose it doesn't really matter much, so if you think it helps to organize I'm good with it 🙂.

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+// Custom types
+// TODO: Remove these and use the version from `basis_search`
+pub type BasisTransforms = Vec<(String, u32, SmallVec<[Param; 3]>, CircuitRep)>;
+pub type MappedTransforms = HashMap<(String, u32), (SmallVec<[Param; 3]>, DAGCircuit)>;
```
IMO, type aliases are generally more helpful when used to break down complex types into logical ideas, rather than as a means of typing less.

For example, this:

```rust
pub type GateIdentifier = (String, u32);
pub type BasisTransformIn = (SmallVec<[Param; 3]>, CircuitRep);
pub type BasisTransformOut = (SmallVec<[Param; 3]>, DAGCircuit);

pub(super) fn py_compose_transforms(
    py: Python,
    basis_transforms: Vec<(GateIdenfitier, BasisTransformIn)>,
    source_basis: HashSet<GateIdentifier>,
    source_dag: &DAGCircuit,
) -> PyResult<HashMap<GateIdentifier, BasisTransformOut>> 
```

gives the reader a lot more information about the call semantics of the function compared to this:

```rust
pub type BasisTransforms = Vec<(String, u32, SmallVec<[Param; 3]>, CircuitRep)>;
pub type MappedTransforms = HashMap<(String, u32), (SmallVec<[Param; 3]>, DAGCircuit)>;

pub(super) fn py_compose_transforms(
    py: Python,
    basis_transforms: BasisTransforms,
    source_basis: HashSet<(String, u32)>,
    source_dag: &DAGCircuit,
) -> PyResult<MappedTransforms> 
```

And, as a general rule of thumb, it's usually best to avoid aliasing data structures like `Vec` and `HashMap` since it hides their otherwise very familiar semantics from the reader (...though, we're guilty of this in a few places, I expect.)

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+        if ob.is_instance(QUANTUM_CIRCUIT.get_bound(ob.py()))? {
+            let data: Bound<PyAny> = ob.getattr("_data")?;
+            let data_downcast: Bound<CircuitData> = data.downcast_into()?;
+            let data_extract: CircuitData = data_downcast.extract()?;
```
I believe you can extract from the `Bound<PyAny>` without first downcasting to save an `isinstance` check.

raynelfss replied:
Yes, this is leftover code from when I was using `Borrow` instead of `extract()`, I will remove the downcast.

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -2861,7 +2861,7 @@ def _format(operand):
     ///     DAGCircuitError: if met with unexpected predecessor/successors
     #[pyo3(signature = (node, input_dag, wires=None, propagate_condition=true))]
-    fn substitute_node_with_dag(
+    pub fn substitute_node_with_dag(
```
Perhaps a strange request, but I think it makes sense for us to rename any PyO3 methods to `py_` when opening them up outside of `DAGCircuit`. The reason being that it'll make it easier for us to notice calls to Python-intended methods from within Rust, which are likely to be ill-performing in contrast to anything proper Rust native.

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+    py: Python,
+    dag: &DAGCircuit,
+    example_gates: Option<Box<HashMap<(String, u32), PackedInstruction>>>,
+) -> PyResult<Box<HashMap<(String, u32), PackedInstruction>>> {
```
The original Python function is _strange_. A more appropriate way to do this in Rust would be to require that the caller allocates the map and provides a mutable reference to it so all recursive calls to the function can keep appending to it.

```suggestion
    example_gates: &mut HashMap<(String, u32), PackedInstruction>,
) -> PyResult<()> {
```

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+) -> PyResult<Box<HashMap<(String, u32), PackedInstruction>>> {
+    let mut example_gates = example_gates.unwrap_or_default();
+    for node in dag.op_nodes(true) {
+        if let Some(NodeType::Operation(op)) = dag.dag().node_weight(node) {
```
I think we now have `NodeType::unwrap_operation` so you can use that here instead of the `if let`.

kevinhartman replied:
Actually, that would come as part of https://github.com/Qiskit/qiskit/pull/13034/files#diff-9d10dd27b7540092c29eb23e0fe83a90b8661fb1f452178b60f0e8fb4fd54d60R85

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+                    panic!("Control flow op can only be an instruction")
+                };
+                let inst_bound = inst.instruction.bind(py);
+                let blocks = inst_bound.getattr("blocks")?;
```
And, you of course now have `Operation::blocks` :)

raynelfss replied:
Yes, I am planning on using it here. However, it does throw a wrench in some of my plans because I can't necessarily verify that `CircuitData` has any calibrations for a specific instruction due to that part of the circuit living entirely on Python.

EDIT: Oh I think I was wrong, I can use it without issue here.

Until then I might keep this as is, but my upcoming implementation of `BasisTranslator` does use `Operation::blocks`,  thank you for that by the way.

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+                let blocks = inst_bound.getattr("blocks")?;
+                for block in blocks.iter()? {
+                    let block_as_dag = circuit_to_dag(py, block?.extract()?, true, None, None)?;
+                    example_gates = get_example_gates(py, &block_as_dag, Some(example_gates))?;
```
Rather than convert the `CircuitData` to a `DAGCircuit`, it'd be more efficient to handle the `CircuitData`s directly.

There's an example of this in `GatesInBasis`. For that, we have handling for the `DAGCircuit` in the root of the function toward the bottom, and define a recursive inner function to then process nested `CircuitData`s.

https://github.com/Qiskit/qiskit/pull/13034/files#diff-6cbc6dd8e1b6e781868b664eee4429814a0cedbe4a4f4d8393413e15c9ae5878R36-R102

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+
+    for (gate_name, gate_num_qubits) in source_basis.iter().cloned() {
+        // Need to grab a gate instance to find num_qubits and num_params.
+        // Can be removed following https://github.com/Qiskit/qiskit-terra/pull/3947 .
```
This issue appears to be closed. I think the comment may be outdated in your implementation, anyhow?

raynelfss replied:
Yes, this was a question I had. Not sure how to tackle it since it seems that issue has been abandoned.

kevinhartman replied:
I would probably just remove the comment, and add a method comment to `get_example_gates` that explains that its purpose is just to get the number of parameters for a gate with a specific name and number of qubits, since that info is only found on an instruction.

(...also, we should probably rename `get_example_gates` to `get_num_params` and change the return type to  `HashMap<(String, u32), usize>`. As it is, we don't use the `PackedInstruction`s coming out of the function (do we?) and they are being cloned)

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+/// Representation of QuantumCircuit which the original circuit object + an
+/// instance of `CircuitData`.
+#[derive(Debug, Clone)]
+pub struct CircuitRep(pub CircuitData);
```
Maybe `CircuitFromPython` could be a nice name, similar to how we have `OperationFromPython`.

kevinhartman commented on qiskit/transpiler/passes/basis/basis_translator.py:
```diff
@@ -597,97 +594,3 @@ def _basis_search(equiv_lib, source_basis, target_basis):
-        dag.apply_operation_back(placeholder_gate, qr, (), check=False)
-        mapped_instrs[gate_name, gate_num_qubits] = placeholder_params, dag
-
-    for gate_name, gate_num_qubits, equiv_params, equiv in basis_transforms:
```
It looks like the documentation is wrong then for `basis_transforms` in the original Python code, no? Since the tuple does not have the number of bits there.

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+            Some(gate.into()),
+        )?;
+        mapped_instructions.insert(
+            (gate_name.clone(), gate_num_qubits),
```
Do we need to clone `gate_name` here, or can we just consume it? I don't see anymore uses after this :)

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,206 @@
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        self.clone().into_py(py)
+    }
+}
```
Are these used?

raynelfss replied:
I don't think so, at least not here. This is leftover code brought over from #12585.

kevinhartman reviewed (commented):
Looks like a good start, thanks for working on this.

Unfortunately, the Python code you're starting with is probably sorely in need of a rewrite. I'm not entirely familiar with it, but it seems riddled with wrong / bad docs and tuple abuse.

I'll need to do a second pass once you've had a look at the comments thus far. I haven't spent enough time reviewing the parameter mapping stuff, specifically.

raynelfss committed:
Merge branch 'main' into rust-compose-trans

raynelfss committed:
Fix: Code review comments
- Remove unused `circuit_to_dag` and `OperationRef` imports.
- Streamline complex types into simpler aliases.
- Rename `CircuitRep` to `CircuitFromPython`.
- Reshape `get_example_gates` to work with `CircuitData` instances during its recursion.
- Remove unnecessary clone of `gate_name` in `compose_transform`.
- Use `mapped_instructions` values when iterating in `compose_transforms`.
- Rename `DAGCircuit::substitute_node_with_dag` to `DAGCircuit::py_*` in rust.
- Adapted `_basis_search` to return a tuple of tuples instead of a 4-tuple.

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,202 @@
+pub(super) fn compose_transforms<'a>(
+    py: Python,
+    basis_transforms: &'a [(GateIdentifier, BasisTransformIn)],
+    source_basis: &'a HashSet<(String, u32)>,
```
```suggestion
    source_basis: &'a HashSet<GateIdentifier>,
```

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,202 @@
+    source_basis: &'a HashSet<(String, u32)>,
+    source_dag: &'a DAGCircuit,
+) -> PyResult<HashMap<GateIdentifier, BasisTransformOut>> {
+    let mut example_gates: HashMap<(String, u32), PackedInstruction> = HashMap::default();
```
```suggestion
    let mut example_gates: HashMap<GateIdentifier, PackedInstruction> = HashMap::default();
```

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,202 @@
+    Ok(mapped_instructions)
+}
+
+fn get_example_gates(
```
I made this comment elsewhere, but I think this can just be `get_num_params` and return `HashMap<GateIdentifier, usize>`, where the value is the number of parameters, since we don't actually use the `PackedInstruction` otherwise, I think.

That would save us from having to clone all of the nodes' `PackedInstruction`s.

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,202 @@
+
+fn get_example_gates(
+    dag: &DAGCircuit,
+    example_gates: &mut HashMap<(String, u32), PackedInstruction>,
```
```suggestion
    example_gates: &mut HashMap<GateIdentifier, PackedInstruction>,
```

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,202 @@
+
+fn get_example_gates_circuit(
+    circuit: &CircuitData,
+    example_gates: &mut HashMap<(String, u32), PackedInstruction>,
```
```suggestion
    example_gates: &mut HashMap<GateIdentifier, PackedInstruction>,
```

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,202 @@
+
+        for ((gate_name, gate_num_qubits), (equiv_params, equiv)) in basis_transforms {
+            for (_, dag) in &mut mapped_instructions.values_mut() {
+                let doomed_nodes = dag
```
I know the name `doomed_nodes` is faithful to the original Python code, but I don't think it's particularly descriptive (even though it is **fun**) for what is happening here.

Maybe `nodes_to_replace` or something to that effect?

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,202 @@
+                        .assign_parameters_from_mapping(py, param_mapping)?;
+                    let replace_dag: DAGCircuit =
+                        DAGCircuit::from_circuit_data(py, replacement.0, true)?;
+                    let op_node = dag.get_node(py, node)?;
+                    dag.py_substitute_node_with_dag(
+                        py,
+                        op_node.bind(py),
+                        &replace_dag,
+                        None,
+                        true,
+                    )?;
```
I imagine this is slower than it ought to be, but it'd be non-trivial to expose a native-Rust `substitute_node_with_dag`. Maybe we can revisit doing so in a future PR.

kevinhartman reviewed (commented):
Just a few more comments. Thanks! :)

raynelfss committed:
Fix: More commments from code review
- Remove stale comment related to #3947
- Rename tuple instances to `GateIdentifier`.
- Rename `doomed_nodes` to `nodes_to_replace`.
- Add docstring for `get_example_gates`.
- Rename `get_example_gates` to `get_gate_num_params`.

Co-authored-by: Kevin Hartman <kevin@hart.mn>

raynelfss committed:
Merge branch 'main' into rust-compose-trans

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,209 @@
+// TODO: Remove these and use the version from `EquivalenceLibrary`
+
+/// Representation of QuantumCircuit which the original circuit object + an
+/// instance of `CircuitData`.
```
It might be good to phrase this to be more clear that it is just used to extract a `CircuitData` from a Python QuantumCircuit, e.g. "Used to extract a Rust `CircuitData` from the Python `QuantumCircuit`."

raynelfss replied:
I will actually add those changes to the original struct which is located in `equivalence.rs`. And remove the duplicate from here.

kevinhartman commented on crates/accelerate/src/basis/basis_translator/compose_transforms.rs:
```diff
@@ -0,0 +1,209 @@
+    source_basis: &'a HashSet<GateIdentifier>,
+    source_dag: &'a DAGCircuit,
+) -> PyResult<HashMap<GateIdentifier, BasisTransformOut>> {
+    let mut example_gates: HashMap<GateIdentifier, usize> = HashMap::default();
```
I think this should be renamed now to something like `gate_param_counts`.

raynelfss committed:
Merge branch 'main' into rust-compose-trans

kevinhartman reviewed (commented):
Last two minor comments. Otherwise, it looks reasonable to me.

raynelfss committed:
Refactor: Rename `example_gates` to `gate_param_counts`
- Remove `CircuitFromPython` and re-use original instance from `EquivalenceLibrary` after #12585 merged.

kevinhartman reviewed (approved):
LGTM, thanks for doing this work!
